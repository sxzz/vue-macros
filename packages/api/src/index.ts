export * from '@vue-macros/common'

export * from './error'
export * from './vue'
export * from './ts'
