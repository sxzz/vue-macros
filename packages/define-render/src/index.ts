import {
  createFilter,
  detectVueVersion,
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import { createUnplugin, type UnpluginInstance } from 'unplugin'
import { transformDefineRender } from './core'

export type Options = BaseOptions & {
  vapor?: boolean
}
export type OptionsResolved = MarkRequired<Options, 'include' | 'version'>

function resolveOptions(options: Options): OptionsResolved {
  const version = options.version || detectVueVersion()
  return {
    include: [
      REGEX_VUE_SFC,
      REGEX_SETUP_SFC,
      /\.(vue|setup\.[cm]?[jt]sx?)\?vue/,
    ],
    version,
    ...options,
  }
}
const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}) => {
    const options = resolveOptions(userOptions)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'post',

      transformInclude(id) {
        return filter(id)
      },

      transform(code, id) {
        return transformDefineRender(code, id, options)
      },
    }
  },
)
export default plugin
