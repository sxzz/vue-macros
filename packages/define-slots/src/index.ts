import {
  createFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { transformDefineSlots } from './core'

export type Options = BaseOptions
export type OptionsResolved = MarkRequired<Options, 'include' | 'version'>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    ...options,
    version,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'pre',
      transformInclude: filter,
      transform: transformDefineSlots,
    }
  },
)
export default plugin
