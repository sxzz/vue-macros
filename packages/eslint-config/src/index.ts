/* eslint perfectionist/sort-objects: "error" */

import type { Rules } from '@sxzz/eslint-config'
import type { Linter } from 'eslint'

const globals: Linter.Globals = {
  $: 'readonly',
  $$: 'readonly',
  $computed: 'readonly',
  $customRef: 'readonly',
  $defineModels: 'readonly',
  $defineProps: 'readonly',
  $definePropsRefs: 'readonly',
  $ref: 'readonly',
  $shallowRef: 'readonly',
  $toRef: 'readonly',
  defineEmit: 'readonly',
  defineModels: 'readonly',
  defineOptions: 'readonly',
  defineProp: 'readonly',
  defineProps: 'readonly',
  defineRender: 'readonly',
  defineSetupComponent: 'readonly',
  defineSlots: 'readonly',
}

const rules: Linter.RulesRecord = {
  'vue/no-export-in-script-setup': 'off', // exportRender / exportExpose / exportProps
  'vue/valid-attribute-name': 'off', // short-vmodel
  'vue/valid-define-props': 'off', // hoistStatic
  'vue/valid-v-bind': 'off', // shortBind + shortVmodel
} satisfies Rules

const config: Linter.Config = {
  languageOptions: {
    globals,
  },
  name: 'vue-macros/rules',
  rules,
}

export default config
