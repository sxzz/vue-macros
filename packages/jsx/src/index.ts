import jsxDirective from '@vue-macros/jsx-directive'
import jsxMacros from '@vue-macros/jsx-macros'

import { generatePluginName } from '#macros' with { type: 'macro' }
import {
  createCombinePlugin,
  type UnpluginCombineInstance,
} from 'unplugin-combine'
import { resolveJSXOptions, type JSXOptions } from './options'

const name = generatePluginName()
const plugin: UnpluginCombineInstance<JSXOptions | undefined> =
  createCombinePlugin<JSXOptions | undefined>((userOptions = {}, meta) => {
    const options = resolveJSXOptions(userOptions)
    const framework = meta.framework!
    const plugins = [
      options.directive
        ? jsxDirective[framework](options.directive)
        : undefined,
      options.macros ? jsxMacros[framework](options.macros) : undefined,
    ].filter((plugin) => !!plugin)

    return {
      name,
      plugins,
    }
  })

export default plugin
