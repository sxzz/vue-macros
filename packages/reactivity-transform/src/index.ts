import {
  createFilter,
  createRollupFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  normalizePath,
  REGEX_NODE_MODULES,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { shouldTransform, transform, transformVueSFC } from './core'
import { helperCode, helperId } from './core/helper'

export type Options = BaseOptions
export type OptionsResolved = MarkRequired<Options, 'include' | 'version'>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.SRC_FILE, FilterFileType.VUE_SFC_WITH_SETUP],
    framework,
  )
  return {
    include,
    exclude: [REGEX_NODE_MODULES],
    ...options,
    version,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)
    const filterSFC = createRollupFilter(
      getFilterPattern(
        [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
        framework,
      ),
    )
    return {
      name,
      enforce: 'pre',

      resolveId(id) {
        if (id === normalizePath(helperId)) return id
      },

      loadInclude(id) {
        return normalizePath(id) === helperId
      },

      load(id) {
        if (normalizePath(id) === helperId) return helperCode
      },

      transformInclude: filter,
      transform(code, id) {
        if (filterSFC(id)) {
          return transformVueSFC(code, id)
        } else if (shouldTransform(code)) {
          return transform(code, {
            filename: id,
            sourceMap: true,
          })
        }
      },
    }
  },
)
export default plugin
