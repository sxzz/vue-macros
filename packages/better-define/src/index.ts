import process from 'node:process'
import { resolveDtsHMR } from '@vue-macros/api'
import {
  createFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { transformBetterDefine } from './core'

export type Options = BaseOptions
export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    isProduction: process.env.NODE_ENV === 'production',
    ...options,
    version,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'pre',

      transformInclude: filter,
      async transform(code, id) {
        try {
          return (
            await transformBetterDefine(code, id, options.isProduction)
          ).match(
            (res) => res,
            (error) => console.warn(error),
          )
        } catch (error: unknown) {
          console.warn(error)
        }
      },

      vite: {
        configResolved(config) {
          options.isProduction ??= config.isProduction
        },

        handleHotUpdate: resolveDtsHMR,
      },
    }
  },
)
export default plugin
