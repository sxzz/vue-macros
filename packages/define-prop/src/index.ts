import process from 'node:process'
import { resolveDtsHMR } from '@vue-macros/api'
import {
  createFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  normalizePath,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { transformDefineProp, type Edition } from './core'
import { helperCode, helperId } from './core/helper'

export interface Options extends BaseOptions {
  edition?: Edition
}

export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction' | 'edition'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    isProduction: process.env.NODE_ENV === 'production',
    edition: 'kevinEdition',
    ...options,
    version,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'pre',

      resolveId(id) {
        if (id === normalizePath(helperId)) return id
      },

      loadInclude(id) {
        return normalizePath(id) === helperId
      },

      load(id) {
        if (normalizePath(id) === helperId) return helperCode
      },

      transformInclude: filter,
      transform(code, id) {
        return transformDefineProp(
          code,
          id,
          options.edition,
          options.isProduction,
        )
      },

      vite: {
        configResolved(config) {
          options.isProduction ??= config.isProduction
        },

        handleHotUpdate: resolveDtsHMR,
      },
    }
  },
)
export default plugin
