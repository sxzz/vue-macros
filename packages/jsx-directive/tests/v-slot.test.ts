import { testFixtures } from '@vue-macros/test-utils'
import { describe } from 'vitest'
import { transformJsxDirective } from '../src/api'

describe('jsx-vue-directive', () => {
  describe('vue 3 v-slot', async () => {
    await testFixtures(
      import.meta.glob<string>('./fixtures/v-slot/index.vue', {
        eager: true,
        query: '?raw',
        import: 'default',
      }),
      (_, id, code) =>
        transformJsxDirective(code, id, {
          version: 3,
          lib: 'vue',
          prefix: 'v-',
        })?.code,
    )
  })

  describe('vue 2.7 v-slot', async () => {
    await testFixtures(
      import.meta.glob<string>('./fixtures/v-slot/index.vue', {
        eager: true,
        query: '?raw',
        import: 'default',
      }),
      (_, id, code) =>
        transformJsxDirective(code, id, {
          version: 2.7,
          lib: 'vue',
          prefix: 'v-',
        })?.code,
    )
  })

  describe('vue/vapor v-slot', async () => {
    await testFixtures(
      import.meta.glob<string>('./fixtures/v-slot/index.vue', {
        eager: true,
        query: '?raw',
        import: 'default',
      }),
      (_, id, code) =>
        transformJsxDirective(code, id, {
          version: 3,
          lib: 'vue/vapor',
          prefix: 'v-',
        })?.code,
    )
  })
})
