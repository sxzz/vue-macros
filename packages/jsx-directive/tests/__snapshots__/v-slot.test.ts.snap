// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jsx-vue-directive > vapor v-slot > ./fixtures/v-slot/index.vue 1`] = `
"<script setup lang="tsx">
import { renderList as __MACROS_renderList } from "vue";
import { unref as __MACROS_unref } from "@vue-macros/jsx-directive/helpers";
import { expectTypeOf } from 'expect-type'
import Child from './child.vue'
import type { FunctionalComponent } from 'vue'

const Comp: FunctionalComponent<
  {},
  {},
  {
    default: (scope: { foo: string }) => any
  }
> = (props, { slots }) => {
  return (
    <Child vSlots={{...Object.fromEntries(__MACROS_renderList(slots, (Slot, slotName) => ([[__MACROS_unref(slotName)], (scope) => <>
        <Slot {...scope} />
      </>,]))),}}
      
    ></Child>
  )
}

let baz = $ref('')
let show = $ref<boolean | undefined>()
defineRender(() => (
  <div>
    <Child  vSlots={{'bottom': (__MACROS_props) => <>
      {__MACROS_props.foo}
      <Child  vSlots={{'default': () => <>default</>,}}></Child>
    </>,}}></Child>

    <Comp  vSlots={{'default': () => <>default</>,}}></Comp>
    <Child  vSlots={{'bottom': (__MACROS_props) => <>{__MACROS_props.foo}</>,}}></Child>

    <Child baz={baz}  vSlots={{'title': (__MACROS_props) => <>
      {expectTypeOf<string>(__MACROS_props.foo)}
    </>,}}></Child>

    <Child vSlots={{...(show) ? {'bottom': (__MACROS_props) => <>
        {__MACROS_props.foo}
      </>,} : null,}}
      
    ></Child>

    <Child<number> vSlots={{...(show) ? {'title': (__MACROS_props) => <>
        {expectTypeOf<number>(__MACROS_props.foo)}
        {show}
      </>,} : (show === false) ? {'center': (__MACROS_props) => <>
        {__MACROS_props.foo}
      </>,} : {'bottom': () => <> </>,},'bot-tom': () => <>bot-tom</>,default: () => <>
      default
      
      
      
      
    </>}}></Child>
  </div>
))
</script>
"
`;

exports[`jsx-vue-directive > vue 2.7 v-slot > ./fixtures/v-slot/index.vue 1`] = `
"<script setup lang="tsx">
import { unref as __MACROS_unref } from "vue";
import { expectTypeOf } from 'expect-type'
import Child from './child.vue'
import type { FunctionalComponent } from 'vue'

const Comp: FunctionalComponent<
  {},
  {},
  {
    default: (scope: { foo: string }) => any
  }
> = (props, { slots }) => {
  return (
    <Child scopedSlots={{...Object.fromEntries(Array.from(slots, (Slot, slotName) => ([[__MACROS_unref(slotName)], (scope) => <span>
        <Slot {...scope} />
      </span>,]))),}}
      
    ></Child>
  )
}

let baz = $ref('')
let show = $ref<boolean | undefined>()
defineRender(() => (
  <div>
    <Child  scopedSlots={{'bottom': ({ foo }) => <span>
      {foo}
      <Child  scopedSlots={{'default': () => <span>default</span>,}}></Child>
    </span>,}}></Child>

    <Comp  scopedSlots={{'default': () => <span>default</span>,}}></Comp>
    <Child  scopedSlots={{'bottom': ({ foo }) => <span>{foo}</span>,}}></Child>

    <Child baz={baz}  scopedSlots={{'title': ({ foo }) => <span>
      {expectTypeOf<string>(foo)}
    </span>,}}></Child>

    <Child scopedSlots={{...(show) ? {'bottom': ({ foo }) => <span>
        {foo}
      </span>,} : null,}}
      
    ></Child>

    <Child<number> scopedSlots={{...(show) ? {'title': ({ foo }) => <span>
        {expectTypeOf<number>(foo)}
        {show}
      </span>,} : (show === false) ? {'center': ({ foo }) => <span>
        {foo}
      </span>,} : {'bottom': () => <span> </span>,},'bot-tom': () => <span>bot-tom</span>,default: () => <span>
      default
      
      
      
      
    </span>}}></Child>
  </div>
))
</script>
"
`;

exports[`jsx-vue-directive > vue 3 v-slot > ./fixtures/v-slot/index.vue 1`] = `
"<script setup lang="tsx">
import { renderList as __MACROS_renderList } from "vue";
import { unref as __MACROS_unref } from "vue";
import { expectTypeOf } from 'expect-type'
import Child from './child.vue'
import type { FunctionalComponent } from 'vue'

const Comp: FunctionalComponent<
  {},
  {},
  {
    default: (scope: { foo: string }) => any
  }
> = (props, { slots }) => {
  return (
    <Child vSlots={{...Object.fromEntries(__MACROS_renderList(slots, (Slot, slotName) => ([[__MACROS_unref(slotName)], (scope) => <>
        <Slot {...scope} />
      </>,]))),}}
      
    ></Child>
  )
}

let baz = $ref('')
let show = $ref<boolean | undefined>()
defineRender(() => (
  <div>
    <Child  vSlots={{'bottom': ({ foo }) => <>
      {foo}
      <Child  vSlots={{'default': () => <>default</>,}}></Child>
    </>,}}></Child>

    <Comp  vSlots={{'default': () => <>default</>,}}></Comp>
    <Child  vSlots={{'bottom': ({ foo }) => <>{foo}</>,}}></Child>

    <Child baz={baz}  vSlots={{'title': ({ foo }) => <>
      {expectTypeOf<string>(foo)}
    </>,}}></Child>

    <Child vSlots={{...(show) ? {'bottom': ({ foo }) => <>
        {foo}
      </>,} : null,}}
      
    ></Child>

    <Child<number> vSlots={{...(show) ? {'title': ({ foo }) => <>
        {expectTypeOf<number>(foo)}
        {show}
      </>,} : (show === false) ? {'center': ({ foo }) => <>
        {foo}
      </>,} : {'bottom': () => <> </>,},'bot-tom': () => <>bot-tom</>,default: () => <>
      default
      
      
      
      
    </>}}></Child>
  </div>
))
</script>
"
`;
